array.h

```diff
@@ -1259,17 +1492,20 @@ auto clamp_dims(const DimA& a, const DimB& b) {
 }
 
 template <class DimsA, class DimsB, size_t... Is>
+NDARRAY_HOST_DEVICE
 auto clamp(const DimsA& a, const DimsB& b, index_sequence<Is...>) {
   return make_shape(clamp_dims(std::get<Is>(a), std::get<Is>(b))...);
 }
 
 // Shuffle a tuple with indices Is...
 template <size_t... Is, class T>
+NDARRAY_HOST_DEVICE
 auto shuffle(const T& t) {
   return std::make_tuple(std::get<Is>(t)...);
 }
 
 // Return where the index I appears in Is...
+// TODO(dsharlet): Use numeric_limits<size_f>::max()?
```

Comment:
This needs some headroom so we can add to it without overflowing. This definitely should have a comment though, so leaving the TODO is fine.